arch/arm/core/elf.c

```diff
@@ -11,42 +12,261 @@
+	if (reloc_type == R_ARM_MOVT_PREL || reloc_type == R_ARM_MOVW_PREL_NC) {
+		*offset -= loc;
+	}
+	if (reloc_type == R_ARM_MOVT_ABS || reloc_type == R_ARM_MOVT_PREL) {
+		*offset >>= 16;
+	}
+
+	tmp &= (MASK_MOV_COND | MASK_MOV_00 | MASK_MOV_I | MASK_MOV_OPCODE | MASK_MOV_RD);
+	tmp |= ((*offset & MASK_MOV_RD) << SHIFT_MOV_RD) |
+	(*offset & (MASK_MOV_OPERAND2));
+
+	*(uint32_t *)loc = OPCODE2ARMMEM(tmp);
 }
 
-static void arm_bl_blx_encode_addend(uintptr_t opaddr, int32_t addend)
+static inline int thm_jumps_decode(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name, int32_t *offset,
+			uint32_t *upper, uint32_t *lower)
 {
-	uint16_t upper = *((uint16_t *)opaddr);
-	uint16_t lower = *(((uint16_t *)opaddr) + 1);
+	int ret;
+	uint32_t j_one, j_two, sign;
+
+	*upper = MEM2THM16OPCODE(*(uint16_t *)loc);
+	*lower = MEM2THM16OPCODE(*(uint16_t *)(loc + 2));
 
-	addend = upper & ARM_BL_BLX_UPPER_S_BIT ? addend >> 1 : addend >> 2;
+	/* sign is bit10 */
+	sign = (*upper >> BIT_THM_BW_S) & 1;
+	j_one = (*lower >> BIT_THM_BL_J1) & 1;
+	j_two = (*lower >> BIT_THM_BL_J2) & 1;
+	*offset = (sign << SHIFT_THM_JUMPS_SIGN) | ((~(j_one ^ sign) & 1) << SHIFT_THM_BL_J1) |
+				((~(j_two ^ sign) & 1) << SHIFT_THM_BL_J2) |
+				((*upper & MASK_THM_BW_IMM10) << SHIFT_THM_BW_IMM10) |
+				((*lower & MASK_THM_BL_IMM11) << 1);
+	*offset = sign_extend(*offset, SHIFT_THM_JUMPS_SIGN);
+	*offset += sym_base_addr - loc;
 
-	upper &= ARM_BL_BLX_HDR_MASK;
-	lower &= ARM_BL_BLX_HDR_MASK;
-	upper |= (addend >> ARM_BL_BLX_ADDEND_SIZE) & ARM_BL_BLX_ADDEND_MASK;
-	lower |= addend & ARM_BL_BLX_ADDEND_MASK;
+	if (*offset <= THM_JUMP_UPPER_BOUNDARY || *offset >= THM_JUMP_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
 
-	*((uint16_t *)opaddr) = upper;
-	*(((uint16_t *)opaddr) + 1) = lower;
+	return ret;
+}
+
+static inline void thm_jumps_reloc(uint32_t loc, int32_t *offset,
+			uint32_t *upper, uint32_t *lower)
+{
+	uint32_t j_one, j_two, sign;
+
+	sign = (*offset >> SHIFT_THM_JUMPS_SIGN) & 1;
+	j_one = sign ^ (~(*offset >> SHIFT_THM_BL_J1) & 1);
+	j_two = sign ^ (~(*offset >> SHIFT_THM_BL_J2) & 1);
+	*upper = (uint16_t)((*upper & MASK_THM_BW_11110) | (sign << BIT_THM_BW_S) |
+				((*offset >> SHIFT_THM_BW_IMM10) & MASK_THM_BW_IMM10));
+	*lower = (uint16_t)((*lower & (MASK_THM_BL_10|MASK_THM_BL_1)) |
+				(j_one << BIT_THM_BL_J1) | (j_two << BIT_THM_BL_J2) |
+				((*offset >> 1) & MASK_THM_BL_IMM11));
+
+	*(uint16_t *)loc = OPCODE2THM16MEM(*upper);
+	*(uint16_t *)(loc + 2) = OPCODE2THM16MEM(*lower);
+}
+
+static int thm_jumps_handler(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	int ret;
+	uint32_t upper, lower;
+
+	ret = thm_jumps_decode(reloc_type, loc, sym_base_addr, sym_name, offset, &upper, &lower);
+	if (!ret) {
+		thm_jumps_reloc(loc, offset, &upper, &lower);
```

Comment:
upper and lower can be passed by value here

Reply:
yes, but what's the benefit?!